source/Agents/GP/GPTrainingLoop.hpp

```diff
@@ -0,0 +1,574 @@
+
+        tinyxml2::XMLDocument doc;
+        tinyxml2::XMLDocument lastGenerationsDoc;
+
+
+        tinyxml2::XMLElement *root = doc.NewElement("GPLoop");
+
+
+        tinyxml2::XMLElement *rootTEMP = lastGenerationsDoc.NewElement("GPLoop");
+
+        std::vector<std::pair<int, int>> sortedAgents = std::vector<std::pair<int, int>>();
+
+//        const std::vector<cse491::GridPosition> STARTPOSITIONS = {cse491::GridPosition(0,0), cse491::GridPosition(4,4), cse491::GridPosition(19,5), cse491::GridPosition(20,1) };
+        const std::vector<cse491::GridPosition> STARTPOSITIONS = {cse491::GridPosition(0,0) };
+
+    public:
+
+        GPTrainingLoop() {
+
+          doc.InsertFirstChild(root);
+          resetMainTagLastGenerations();
+
+        }
+
+        void resetMainTagLastGenerations() {
+          rootTEMP = lastGenerationsDoc.NewElement("GPLoop");
+          lastGenerationsDoc.InsertFirstChild(rootTEMP);
+        }
+
+        /**
+         * @brief Initialize the training loop with a number of environments and agents per environment.
+         * @param numArenas
+         * @param NumAgentsForArena
+         */
+        void initialize(size_t numArenas = 5, size_t NumAgentsForArena = 100) {
+
+
+          for (size_t i = 0; i < numArenas; ++i) {
+            // instantiate a new environment
+            environments.emplace_back(new EnvironmentType());
+
+
+            agents.emplace_back(std::vector<cowboys::GPAgent_ *>());
+            TEMPinitialAgentPositions.emplace_back(std::vector<cse491::GridPosition>());
+            for (size_t j = 0; j < NumAgentsForArena; ++j) {
+
+              cowboys::GPAgent_ &addedAgent = static_cast<cowboys::GPAgent_ &>(environments[i]->template AddAgent<AgentType>(
+                      "Agent " + std::to_string(j)));
+              addedAgent.SetPosition(0, 0);
+              cse491::GridPosition position = addedAgent.GetPosition();
+
+              TEMPinitialAgentPositions[i].emplace_back(position);
+
+              agents[i].emplace_back(&addedAgent);
+
+            }
+
+          }
+
+
+
+        }
+
+
+        double simpleFitnessFunction(cse491::AgentBase &agent, cse491::GridPosition startPosition) {
+
+          cse491::GridPosition currentPosition = agent.GetPosition();
+
+          // Eucledian distance
+          double distance = std::sqrt(std::pow(currentPosition.GetX() - startPosition.GetX(), 2) +
+                                      std::pow(currentPosition.GetY() - startPosition.GetY(), 2));
+
+          // Agent complexity, temporarily doing this in a bad way
+          double complexity = 0;
+          if (auto *cgp = dynamic_cast<CGPAgent *>(&agent)) {
+            auto genotype = cgp->GetGenotype();
+            complexity = std::divides<double>()(genotype.GetNumConnections(), genotype.GetNumPossibleConnections());
+          }
+
+          return distance - complexity;
+        }
+
+
+        void run(size_t numGenerations,
```

Comment:
This function is massive and tough to look over. Could it be broken up any more?